### PR TITLE
generate semi-permanent content hashes for OA rows

### PR DESCRIPTION
**Pull request by missinglink**
Following on from a discussion with the OA team, we will begin generating our own content hashes for each row of the CSV file.

The original intention of the `HASH` field was to provide a semi-stable identifier for each row in order to:
- provide an addressable identifier for each row
- facilitate deduplication between imports (such as duplicates in the `us/or/portland_metro` and `us/or/portland` files).
- allow for blacklisting of records using their GID

It turns out that the existing `HASH` column generated by the OA team is [seeded with a random number](https://github.com/openaddresses/machine/blob/b08a622669af7d896739f87891c13ca8b527b5b9/openaddr/conform.py#L1003), so even if the underlying data remains the same, the hash value will change with each rebuild of the OA file.

eg:
```bash
grep 4012 565265_portland_grep.txt 677463_portland_grep.txt
565265_portland_grep.txt:-122.6474699,45.4909803,4012,SE 17TH AVE,,PORTLAND,MULTNOMAH,OR,97202,,9fc860391a6ce4df
677463_portland_grep.txt:-122.6474699,45.4909803,4012,SE 17TH AVE,,PORTLAND,MULTNOMAH,OR,97202,,18eb1f0f484d2e7e
```

This PR puts the responsibility of hash generation on us, it's pretty simple and open for feedback.

### CODE REVIEW

File: lib/streams/contentHashStream.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+const crypto = require('crypto');
+const through2 = require('through2');
+
+/*
+ * create a stream that generates a content-hash for each row
+ */
+
+function createContentHashStream() {
+  return through2.obj(function (record, enc, next) {
+    record.HASH = hash(record);
```
Comment (orangejulius) [inline lib/streams/contentHashStream.js]: so this would basically overwrite the existing `HASH` value from OA, while it continues to exist, right?

If so I really like it. It means the rest of the code can continue to use the value exactly as it expects :)

Comment (missinglink) [reply lib/streams/contentHashStream.js]: yep, I figured it would be easier that way :)

**Comment by missinglink** [comment]
I'd like to give this a little more consideration before merging, simply because, although we can change it to fix issues later, the hash will change, kinda defeating the purpose.

One thing worth considering is what changed in the source data might trigger a hash change.

With the current implementation, the following things would cause the hash to change despite not having any real impact on Pelias or how it functions:
- A column is added/removed/renamed by the OA team from all CSV files (for whatever reason)
- Metadata which we don't import (city,district,region) is added/removed/changed
- Postcode data is added/removed/changed
- Floating-point numbers change due to minor projection changes (different hardware/software) or due to rounding.

Here's an example of one row from an OA CSV file:

```bash
LON,LAT,NUMBER,STREET,UNIT,CITY,DISTRICT,REGION,POSTCODE,ID,HASH
173.0278602,-40.8618603,91,Awaroa Inlet,,,,,,1103299,c0477cde749a5b18
```

We could generate a hash which was considerably more stable in the long-term by doing something like this:

```javascript
h = hash()
h.update( normalizeFloat( LON ) )
h.update( normalizeFloat( LAT ) )
h.update( normalizeString( STREET ) )
h.update( normalizeString( NUMBER ) )
h.update( normalizeString( UNIT ) )
```

Note that this excludes postcode which I'm considering as `metadata` here, so it's free to change without affecting the hash.

The general idea is that:
- `hash( LON, LAT )` is generally enough to ensure uniqueness
- `hash( STREET, NUMBER )` is useful for places like the apt building I live in, which is on the corner of two streets and has two enterances, each with a different street and housenumber.
- `hash( UNIT )` to distinguish different individual apartments withing the building/campus

**Comment by missinglink** [comment]
Added https://github.com/pelias/openaddresses/pull/442/commits/f4f76bf04808eaa98024a2b494bc06cb2a227eb8 to action my concerns in the comment above.

**Comment by missinglink** [comment]
@pelias/contributors any comments on this before merging?
tl;dr it changes the way GIDs are generated for OA records to make them stable between builds.

**Comment by Joxit** [comment]
Your choice for creating the hash only with columns used by pelias is great.

I did a small import yesterday and it looks good :+1:
